### PR TITLE
fix issue 4153, add progress bar to outlier

### PR DESCRIPTION
**Pull request by chunibyo-wly**
#4153 
add progress bar output to two remove outlier function.

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4246)

### CODE REVIEW

**Comment by chunibyo-wly** [comment]
@yxlao 
Hello, thanks for your work.
And I want to change Python API, I tried to imitate the original code and modify it, but I face this style error.
Could you please help me fix this error?

**Comment by yxlao** [comment]
@chunibyo-wly you'll need `yapf` and `clang-format`. See http://www.open3d.org/docs/latest/contribute/styleguide.html

**Comment by chunibyo-wly** [comment]
Thanks for your reply, I' ll check this

**Comment by chunibyo-wly** [comment]
save problem 😞

**Comment by yxlao** [comment]
@chunibyo-wly you're good, `clang-format` and `yapf` worked as expected. 
The warnings came from "Codacy Static Code Analysis", which we can ignore in this case.

File: cpp/open3d/geometry/PointCloud.cpp
Comment (yxlao) [inline cpp/open3d/geometry/PointCloud.cpp:522]: Let's do it properly. We'll need to handle concurrency inside a parallel for loop. See the discussions in https://stackoverflow.com/q/28050669/1255535.

Here are my suggestions:
1. Wait till #4254 is merged. This will move the progress bar to a standalone file so it is easier to work with.
2. I would sub-class `ProgressBar` to be `OMPProgressBar` to make our intention explicit.
3. In `ParallelProgressBar::operator++()`, 1) check if you're inside an OpenMP for loop, 2) check how many threads are used, 3) update your percentage only when thread_id == 0 (taking the idea from https://stackoverflow.com/a/44555438/1255535).
4. Add a unit test (`cpp/tests/utility/ProgressBar.cpp`) for a simple for loop and the new OMP progress bar.

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/open3d/geometry/PointCloud.cpp
Comment (chunibyo-wly) [reply cpp/open3d/geometry/PointCloud.cpp:522]: thanks for your reply, I will fix this later.
May I ask if your third suggestion is to express `OMPProgressBar::operator++()`?

Comment (yxlao) [reply cpp/open3d/geometry/PointCloud.cpp:522]: Yes, `OMPProgressBar::operator++()`. I've updated my comment.

Comment (yxlao) [reply cpp/open3d/geometry/PointCloud.cpp:522]: FYI, @4254 is merged so you may start to refactor the code

Comment (chunibyo-wly) [reply cpp/open3d/geometry/PointCloud.cpp:522]: thanks

**Comment by chunibyo-wly** [comment]
image
hello, I meet a ci problem, This check keeps going round and round and I can't see the error log, could you please re trigger this one? Thank you

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 2 of 6 files at r1, 3 of 3 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4246)* status: all files reviewed, 11 unresolved discussions (waiting on @chunibyo-wly)

___
*[cpp/open3d/utility/OMPProgressBar.h, line 31 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocEJZy2px_pbv9yIXq:-MocEJZy2px_pbv9yIXr:bnsaww) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/open3d/utility/OMPProgressBar.h#L31)):*
> ```objc
> #ifdef _OPENMP
> /// Multi-threading - yay!
> #include <omp.h>
> ```

Avoid third-party header in `.h` files when possible. Move to the `.cpp` file.
___
*[cpp/open3d/utility/OMPProgressBar.h, line 34 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocDxBkEoOEaC4L6Q3u:-MocDxBkEoOEaC4L6Q3v:bluimer) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/open3d/utility/OMPProgressBar.h#L34)):*
> ```objc
> #else
> /// Macros used to disguise the fact that we do not have multithreading enabled.
> #define omp_get_thread_num() 0
> ```

Let's not overwrite this function. 
Instead, disable a section of code such that it never calls `omp_get_num_threads()`, when OpenMP is turned off.
___
*[cpp/open3d/utility/OMPProgressBar.h, line 43 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocET-1DqwEUvt8RXId:-MocET-1DqwEUvt8RXIe:bl8pnf6) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/open3d/utility/OMPProgressBar.h#L43)):*
> ```objc
> namespace utility {
> 
> class OMPProgressBar : public ProgressBar {
> ```

Move this class and to the `ProgressBar.h` (same with the .cpp file). We don't need to create two new files.
___
*[cpp/open3d/utility/OMPProgressBar.cpp, line 27 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocFoS98B1EBBUTZmI2:-MocFoS98B1EBBUTZmI3:b-1lgyrj) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/open3d/utility/OMPProgressBar.cpp#L27)):*
> ```cpp
> // ----------------------------------------------------------------------------
> 
> #include "OMPProgressBar.h"
> ```

nit: Include the full path. See other files for examples.
___
*[cpp/open3d/utility/OMPProgressBar.cpp, line 50 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocDlHN0qXDJP7UR8LR:-MocDlHOAhtxirAG6xXW:b-tkjn4v) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/open3d/utility/OMPProgressBar.cpp#L50)):*
> ```cpp
> 
> }  // namespace utility
> }  // namespace open3d
> ```

nit: Add an extra line below. 
___
*[cpp/open3d/utility/ProgressBar.h, line 48 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocEkwR5Qf7THGUNRcJ:-MocEkwS-4Ln0MlIGCN1:b-wxqmpf) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/open3d/utility/ProgressBar.h#L48)):*
> ```objc
>     void SetCurrentCount(size_t n);
> 
>     void UpdateCurrentCount(size_t n);
> ```

Do we still need this function? If it is not used externally, we can remove this.
___
*[cpp/open3d/utility/ProgressBar.h, line 50 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocEuBaCdO9pSXZhH5Y:-MocEuBaCdO9pSXZhH5Z:bng8plh) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/open3d/utility/ProgressBar.h#L50)):*
> ```objc
>     void UpdateCurrentCount(size_t n);
> 
>     size_t GetCurrentCount() const;
> ```

Same. Do we still need this function? If it is not used externally, we can remove this.
___
*[cpp/pybind/geometry/pointcloud.cpp, line 120 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocF8ix0l6bynb0mAUM:-MocF8ix0l6bynb0mAUN:b-me8dtc) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/pybind/geometry/pointcloud.cpp#L120)):*
> ```cpp
>                  "Function to remove points that have less than nb_points"
>                  " in a given sphere of a given radius",
>                  "nb_points"_a, "radius"_a, "print_progress"_a = false)
> ```

Search for `remove_statistical_outlier` in the same file. You'll find `docstring::ClassMethodDocInject`. Inject the docstring for `print_progress`. 

You can also search for `print_progress` in the same file to see how the `print_progress`'s docstring was injected.
___
*[cpp/pybind/geometry/pointcloud.cpp, line 125 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocFMtg-r-vaCS45H-Q:-MocFMtg-r-vaCS45H-R:b-me8dtc) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/pybind/geometry/pointcloud.cpp#L125)):*
> ```cpp
>                  "Function to remove points that are further away from their "
>                  "neighbors in average",
>                  "nb_neighbors"_a, "std_ratio"_a, "print_progress"_a = false)
> ```

Search for `remove_statistical_outlier` in the same file. You'll find `docstring::ClassMethodDocInject`. Inject the docstring for `print_progress`. 

You can also search for `print_progress` in the same file to see how the `print_progress`'s docstring was injected.
___
*[cpp/tests/utility/ProgressBar.cpp, line 35 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocFiFp0_qBmKx7ppZX:-MocFiFp0_qBmKx7ppZY:bkkmhnu) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/tests/utility/ProgressBar.cpp#L35)):*
> ```cpp
> namespace open3d {
> namespace tests {
> TEST(ProgressBar, ProgressBar) {
> ```

nit: add new line above
___
*[cpp/tests/utility/ProgressBar.cpp, line 59 at r4](https://reviewable.io/reviews/isl-org/open3d/4246#-MocFjp-8o3Orc1IaW8U:-MocFjp03yhf22LrpxI8:b7oicm4) ([raw file](https://github.com/isl-org/open3d/blob/8db5872fe2f2c8a0db340c906111481c7d765ae1/cpp/tests/utility/ProgressBar.cpp#L59)):*
> ```cpp
> 
> }  // namespace tests
> }  // namespace open3d
> ```

nit: add new line below. You'll see a red symbol in github without a new line.

**Review by yxlao — APPROVED** [review]
Review issues have been addressed. LGTM after CI.

> Reviewed 7 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4246#-:-MomCJMD2sX7XIC7hXCK:bs9eqpq)* status: 11 of 13 files reviewed, 5 unresolved discussions (waiting on @chunibyo-wly and @yxlao)

**Pull request merged**
